### ISSUE TITLE
Async Timeout policies do not cancel sync-blocking Thread.Sleep(...)

### ISSUE THREAD

**Issue by shagilt** [App-vNext/Polly#629]
**Summary: What are you wanting to achieve?**
This is a UnitTest code. I have a handler with high latency(which has high latency than configured timeout). But the timeout(client.Timeout) or Polly timeout policies are not in this scenario.

Any idea what is wrong here?

**What code or approach do you have so far?**  
```
public static IServiceCollection AddTestHttpClientWithRetry(this IServiceCollection serviceCollections, string clientName)
        {
            serviceCollections.AddTransient<TestHandler>();
            serviceCollections
                .AddTransient<FirstHandler>();
            serviceCollections.AddHttpClient(clientName, client =>
            {
                client.Timeout = TimeSpan.FromMilliseconds(10000);
            })
                .AddDefaultRetrySettings()
                .AddHttpMessageHandler<FirstHandler>()
                .AddPolicyHandler(SelectTimoutPolicy)
                .AddHttpMessageHandler<TestHandler>();// <--Take more than 30 seconds
            return serviceCollections;
        }

public static IAsyncPolicy<HttpResponseMessage> SelectTimoutPolicy(IServiceProvider sp, HttpRequestMessage httpRequestMessage)
        {
            return Policy.TimeoutAsync<HttpResponseMessage>(
                seconds: 30,
                timeoutStrategy: TimeoutStrategy.Optimistic);
        }

internal class TestHandler : DelegatingHandler
    {
        private readonly Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> _handlerFunc;

        public TestHandler()
        {
            _handlerFunc = (r, c) => WithHighLatency();
        }

        public async Task<HttpResponseMessage> WithHighLatency()
        {
            return await Task.Factory.StartNew(
                () =>
                {
                    Thread.Sleep(20000);
                    return new HttpResponseMessage(HttpStatusCode.OK);
                });
        }
    }

```

**Comment by reisenberger** [comment]
Hi @shagilt .  

+ Async timeout policy with `TimeoutStrategy.Optimistic` expects delegates to honour [co-operative cancellation](https://msdn.microsoft.com/en-us/library/dd997364.aspx) as [documented in the wiki](https://github.com/App-vNext/Polly/wiki/Timeout#optimistic-timeout). The `Thread.Sleep(...)` in the test code does not.
+ Async timeout policy with `TimeoutStrategy.Pessimistic` (to cover this case) is intentionally not designed to time out _purely synchronous_ delegates (more on this also [in the wiki](https://github.com/App-vNext/Polly/wiki/Timeout#additional-notes-pessimistic-timeout-for-async-executions)), as to do so would add unnecessary indirection and performance penalty to the majority well-behaved async case.  So, in addition, switching your policy to async timeout with `TimeoutStrategy.Pessimistic` trying to time out `Thread.Sleep(...)`, would also not time out the delegate.

Longer explanation on above points: https://github.com/App-vNext/Polly/issues/623#issuecomment-475367113, https://github.com/App-vNext/Polly/issues/340#issuecomment-340913315, https://github.com/App-vNext/Polly/issues/318#issuecomment-331139752

The bottom line is: `Thread.Sleep(...)` isn't a good simulation of how calls through `HttpClient` behave, as operations through `HttpClient` do honour `CancellationToken`.  Change to `await Task.Delay(20000, cancellationToken)`, for a realistic simulation.

**Comment by shagilt** [comment]
My bad. Thanks @reisenberger I should have reviewed that part.

**Comment by reisenberger** [comment]
No worries @shagilt . Thanks for everything you are around Polly!  Closing this out as assuming the previous explanation covers, but let us know if you need anything else on this one.
